File: maze_game.js
class Item {
    constructor(name, description, effectType = null, effectValue = 0) {
        this.name = name;
        this.description = description;
        this.effectType = effectType;
        this.effectValue = effectValue;
    }
}

class Room {
    constructor(name, description) {
        this.name = name;
        this.description = description;
        this.exits = {};
        this.items = [];
        this.monster = null;
        this.treasure = false;
        this.puzzleSolved = false;
        this.puzzleQuestion = null;
        this.puzzleAnswer = null;
    }
}

class Monster {
    constructor(name, hp, attack) {
        this.name = name;
        this.hp = hp;
        this.attack = attack;
    }
}

class Player {
    constructor(name) {
        this.name = name;
        this.hp = 100;
        this.maxHp = 100;
        this.attack = 10;
        this.defense = 5;
        this.inventory = [];
        this.treasuresFound = 0;
    }

    takeDamage(damage) {
        const actualDamage = Math.max(1, damage - this.defense);
        this.hp = Math.max(0, this.hp - actualDamage);
        return actualDamage;
    }

    heal(amount) {
        this.hp = Math.min(this.maxHp, this.hp + amount);
    }

    addItem(item) {
        this.inventory.push(item);
        if (item.effectType === "attack") {
            this.attack += item.effectValue;
        } else if (item.effectType === "defense") {
            this.defense += item.effectValue;
        } else if (item.effectType === "heal") {
            this.heal(item.effectValue);
        }
    }
}

class MazeGame {
    constructor() {
        this.player = null;
        this.currentRoom = null;
        this.rooms = {};
        this.gameOver = false;
        this.setupGame();
    }

    setupGame() {
        // 创建房间
        const entrance = new Room("入口大厅", "你站在一个古老的迷宫入口，四周墙壁上刻着神秘的符文。");
        entrance.exits = { east: "神秘图书馆", north: "战斗大厅" };
        entrance.items = [new Item("小刀", "一把锋利的小刀，可以用来战斗。", "attack", 3)];

        const library = new Room("神秘图书馆", "这里堆满了古老的书籍，空气中弥漫着墨水的味道。");
        library.exits = { west: "入口大厅", north: "宝藏室" };
        library.items = [new Item("魔法书", "一本古老的魔法书，可以增加你的攻击力。", "attack", 5)];

        const battleRoom = new Room("战斗大厅", "这个大厅中央有一只凶猛的狼人！它正盯着你。");
        battleRoom.exits = { south: "入口大厅", east: "花园" };
        battleRoom.monster = new Monster("狼人", 30, 8);

        const treasureRoom = new Room("宝藏室", "这个房间闪闪发光，但被一道魔法门锁住了。你需要解决谜题才能进入。");
        treasureRoom.exits = { south: "神秘图书馆" };
        treasureRoom.treasure = true;
        treasureRoom.puzzleQuestion = "什么东西越洗越脏？";
        treasureRoom.puzzleAnswer = "2";
        treasureRoom.items = [new Item("钥匙", "一把神秘的钥匙，可以打开某些门。")];

        const garden = new Room("花园", "一个美丽的花园，充满了奇异的植物和花朵。");
        garden.exits = { west: "战斗大厅", east: "地下密室" };
        garden.items = [new Item("草药", "可以恢复少量生命值的草药。", "heal", 20)];

        const dungeon = new Room("地下密室", "阴暗潮湿的密室，你听到了奇怪的声音。");
        dungeon.exits = { west: "花园", north: "最终宝藏室" };
        dungeon.monster = new Monster("骷髅战士", 25, 6);

        const finalTreasure = new Room("最终宝藏室", "这是迷宫的最终宝藏室！巨大的宝藏就在眼前！");
        finalTreasure.exits = { south: "地下密室" };
        finalTreasure.treasure = true;

        this.rooms = {
            "入口大厅": entrance,
            "神秘图书馆": library,
            "战斗大厅": battleRoom,
            "宝藏室": treasureRoom,
            "花园": garden,
            "地下密室": dungeon,
            "最终宝藏室": finalTreasure
        };

        this.currentRoom = entrance;
    }

    startGame() {
        console.log("欢迎来到神秘迷宫寻宝游戏！");
        console.log("=".repeat(50));
        console.log("你将扮演一个勇敢的冒险者，探索神秘的迷宫并寻找宝藏！");
        
        this.player = new Player("冒险者");
        console.log(`你好，勇敢的 ${this.player.name}！`);
        console.log("你的目标是探索迷宫，收集宝藏，并安全地逃离。");
        console.log(`你拥有 ${this.player.hp} 点生命值。`);
        console.log("=".repeat(50));
        
        this.gameLoop();
    }

    showRoomInfo() {
        const room = this.currentRoom;
        console.log(`\n=== ${room.name} ===`);
        console.log(room.description);
        
        if (Object.keys(room.exits).length > 0) {
            console.log(`出口方向: ${Object.keys(room.exits).join(', ')}`);
        }
        
        if (room.items.length > 0) {
            console.log("房间内的物品:");
            room.items.forEach(item => {
                console.log(`- ${item.name}: ${item.description}`);
            });
        }
        
        if (room.monster) {
            console.log(`危险！你发现了 ${room.monster.name}！`);
            console.log(`它拥有 ${room.monster.hp} 点生命值，攻击力为 ${room.monster.attack}。`);
        }
        
        if (room.treasure && !room.puzzleSolved) {
            console.log("这里有一道谜题需要解决才能获得宝藏。");
        }
        
        if (room.treasure && room.puzzleSolved) {
            console.log("恭喜！你发现了宝藏！");
        }
    }

    showPlayerStatus() {
        console.log(`\n=== ${this.player.name} 的状态 ===`);
        console.log(`生命值: ${this.player.hp}/${this.player.maxHp}`);
        console.log(`攻击力: ${this.player.attack}`);
        console.log(`防御力: ${this.player.defense}`);
        console.log(`收集的宝藏: ${this.player.treasuresFound}`);
        
        if (this.player.inventory.length > 0) {
            console.log("背包中的物品:");
            this.player.inventory.forEach(item => {
                console.log(`- ${item.name}`);
            });
        }
    }

    movePlayer(direction) {
        if (direction in this.currentRoom.exits) {
            const newRoomName = this.currentRoom.exits[direction];
            this.currentRoom = this.rooms[newRoomName];
            console.log(`你移动到了 ${this.currentRoom.name}。`);
        } else {
            console.log("那个方向没有出口！");
        }
    }

    pickItem() {
        const room = this.currentRoom;
        if (room.items.length === 0) {
            console.log("这个房间里没有物品可以拾取。");
            return;
        }
        
        console.log("你想拾取哪个物品？");
        room.items.forEach((item, index) => {
            console.log(`${index + 1}. ${item.name}`);
        });
        
        const choice = parseInt(prompt("请输入选择: ")) - 1;
        if (choice >= 0 && choice < room.items.length) {
            const item = room.items.splice(choice, 1)[0];
            this.player.addItem(item);
            console.log(`你拾取了 ${item.name}。`);
            
            if (item.effectType === "attack") {
                console.log(`你的攻击力增加了 ${item.effectValue} 点！`);
            } else if (item.effectType === "defense") {
                console.log(`你的防御力增加了 ${item.effectValue} 点！`);
            } else if (item.effectType === "heal") {
                console.log(`你恢复了 ${item.effectValue} 点生命值！`);
            }
        } else {
            console.log("无效的选择。");
        }
    }

    combat() {
        const room = this.currentRoom;
        if (!room.monster) {
            console.log("这里没有怪物可以战斗。");
            return;
        }
        
        const monster = room.monster;
        console.log(`战斗开始！你与 ${monster.name} 展开了战斗！`);
        
        while (this.player.hp > 0 && monster.hp > 0) {
            // 玩家攻击
            const playerDamage = Math.floor(Math.random() * this.player.attack) + 1;
            monster.hp -= playerDamage;
            console.log(`你对 ${monster.name} 造成了 ${playerDamage} 点伤害！`);
            
            if (monster.hp <= 0) {
                console.log(`你击败了 ${monster.name}！`);
                room.monster = null;
                break;
            }
            
            // 怪物攻击
            const monsterDamage = Math.max(1, monster.attack - this.player.defense);
            const actualDamage = this.player.takeDamage(monsterDamage);
            console.log(`${monster.name} 对你造成了 ${actualDamage} 点伤害！`);
            
            if (this.player.hp <= 0) {
                console.log(`你被 ${monster.name} 击败了...`);
                this.gameOver = true;
                return;
            }
            
            // 添加延迟让玩家能看到战斗过程
            setTimeout(() => {}, 1000);
        }
    }

    solvePuzzle() {
        const room = this.currentRoom;
        if (!room.treasure || room.puzzleSolved) {
            console.log("这里没有谜题需要解决。");
            return;
        }
        
        console.log("你发现了一个神秘的谜题...");
        console.log(`谜题：${room.puzzleQuestion}`);
        console.log("1. 水  2. 手  3. 布  4. 答案是2");
        
        const answer = prompt("请输入你的选择 (1-4): ");
        
        if (answer === room.puzzleAnswer) {
            console.log("恭喜！谜题解决了！你获得了宝藏！");
            room.puzzleSolved = true;
            this.player.treasuresFound += 1;
        } else {
            console.log("答案错误，再想想吧！");
        }
    }

    showHelp() {
        console.log("\n=== 游戏帮助 ===");
        console.log("look - 查看当前房间信息");
        console.log("status - 查看你的状态");
        console.log("move [方向] - 移动到指定方向 (north, south, east, west)");
        console.log("take - 拾取物品");
        console.log("fight - 战斗");
        console.log("puzzle - 解决谜题");
        console.log("inventory - 查看背包");
        console.log("help - 显示帮助");
        console.log("quit - 退出游戏");
    }

    gameLoop() {
        console.log("\n游戏开始！输入 'help' 查看可用命令。");
        
        while (!this.gameOver) {
            const command = prompt("\n> ").trim().toLowerCase();
            
            if (command === "quit") {
                console.log("感谢游玩神秘迷宫寻宝游戏！");
                break;
            } else if (command === "help") {
                this.showHelp();
            } else if (command === "look") {
                this.showRoomInfo();
            } else if (command === "status") {
                this.showPlayerStatus();
            } else if (command === "inventory") {
                this.showPlayerStatus();
            } else if (command === "take") {
                this.pickItem();
            } else if (command === "fight") {
                this.combat();
            } else if (command === "puzzle") {
                this.solvePuzzle();
            } else if (command.startsWith("move ")) {
                const direction = command.substring(5);
                this.movePlayer(direction);
            } else {
                console.log("无效的命令。输入 'help' 查看可用命令。");
            }
            
            // 检查游戏结束条件
            if (this.player.treasuresFound >= 2) {
                console.log("\n恭喜！你收集了足够的宝藏，成功逃离了迷宫！");
                console.log("你是一个真正的冒险者！");
                break;
            }
            
            if (this.player.hp <= 0) {
                console.log("\n游戏结束！你被击败了...");
                break;
            }
        }
    }
}

// 为了在Node.js环境中运行，我们需要重写prompt函数
function prompt(question) {
    const readline = require('readline');
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    
    return new Promise((resolve) => {
        rl.question(question, (answer) => {
            rl.close();
            resolve(answer);
        });
    });
}

// 如果直接运行这个文件，启动游戏
if (require.main === module) {
    // 由于Node.js的异步特性，我们需要调整游戏循环
    const game = new MazeGame();
    
    // 创建一个简单的命令行界面
    const readline = require('readline');
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    
    console.log("欢迎来到神秘迷宫寻宝游戏！");
    console.log("=".repeat(50));
    console.log("你将扮演一个勇敢的冒险者，探索神秘的迷宫并寻找宝藏！");
    
    game.player = new Player("冒险者");
    console.log(`你好，勇敢的 ${game.player.name}！`);
    console.log("你的目标是探索迷宫，收集宝藏，并安全地逃离。");
    console.log(`你拥有 ${game.player.hp} 点生命值。`);
    console.log("=".repeat(50));
    
    console.log("\n游戏开始！输入 'help' 查看可用命令。");
    
    function processCommand() {
        rl.question("\n> ", (input) => {
            const command = input.trim().toLowerCase();
            
            if (command === "quit") {
                console.log("感谢游玩神秘迷宫寻宝游戏！");
                rl.close();
                return;
            } else if (command === "help") {
                game.showHelp();
                processCommand();
            } else if (command === "look") {
                game.showRoomInfo();
                processCommand();
            } else if (command === "status") {
                game.showPlayerStatus();
                processCommand();
            } else if (command === "inventory") {
                game.showPlayerStatus();
                processCommand();
            } else if (command === "take") {
                game.pickItem();
                processCommand();
            } else if (command === "fight") {
                game.combat();
                processCommand();
            } else if (command === "puzzle") {
                game.solvePuzzle();
                processCommand();
            } else if (command.startsWith("move ")) {
                const direction = command.substring(5);
                game.movePlayer(direction);
                processCommand();
            } else {
                console.log("无效的命令。输入 'help' 查看可用命令。");
                processCommand();
            }
            
            // 检查游戏结束条件
            if (game.player.treasuresFound >= 2) {
                console.log("\n恭喜！你收集了足够的宝藏，成功逃离了迷宫！");
                console.log("你是一个真正的冒险者！");
                rl.close();
                return;
            }
            
            if (game.player.hp <= 0) {
                console.log("\n游戏结束！你被击败了...");
                rl.close();
                return;
            }
        });
    }
    
    processCommand();
}

module.exports = MazeGame;

File: maze_game.py
import random
import time
import os

class Item:
    def __init__(self, name, description, effect_type=None, effect_value=0):
        self.name = name
        self.description = description
        self.effect_type = effect_type
        self.effect_value = effect_value

class Room:
    def __init__(self, name, description):
        self.name = name
        self.description = description
        self.exits = {}
        self.items = []
        self.monster = None
        self.treasure = False
        self.puzzle_solved = False
        self.puzzle_question = None
        self.puzzle_answer = None

class Monster:
    def __init__(self, name, hp, attack):
        self.name = name
        self.hp = hp
        self.attack = attack

class Player:
    def __init__(self, name):
        self.name = name
        self.hp = 100
        self.max_hp = 100
        self.attack = 10
        self.defense = 5
        self.inventory = []
        self.treasures_found = 0
        
    def take_damage(self, damage):
        actual_damage = max(1, damage - self.defense)
        self.hp = max(0, self.hp - actual_damage)
        return actual_damage
        
    def heal(self, amount):
        self.hp = min(self.max_hp, self.hp + amount)
        
    def add_item(self, item):
        self.inventory.append(item)
        if item.effect_type == "attack":
            self.attack += item.effect_value
        elif item.effect_type == "defense":
            self.defense += item.effect_value
        elif item.effect_type == "heal":
            self.heal(item.effect_value)

class MazeGame:
    def __init__(self):
        self.player = None
        self.current_room = None
        self.rooms = {}
        self.game_over = False
        self.setup_game()
        
    def setup_game(self):
        # 创建房间
        entrance = Room("入口大厅", "你站在一个古老的迷宫入口，四周墙壁上刻着神秘的符文。")
        entrance.exits = {"east": "神秘图书馆", "north": "战斗大厅"}
        entrance.items = [Item("小刀", "一把锋利的小刀，可以用来战斗。", "attack", 3)]
        
        library = Room("神秘图书馆", "这里堆满了古老的书籍，空气中弥漫着墨水的味道。")
        library.exits = {"west": "入口大厅", "north": "宝藏室"}
        library.items = [Item("魔法书", "一本古老的魔法书，可以增加你的攻击力。", "attack", 5)]
        
        battle_room = Room("战斗大厅", "这个大厅中央有一只凶猛的狼人！它正盯着你。")
        battle_room.exits = {"south": "入口大厅", "east": "花园"}
        battle_room.monster = Monster("狼人", 30, 8)
        
        treasure_room = Room("宝藏室", "这个房间闪闪发光，但被一道魔法门锁住了。你需要解决谜题才能进入。")
        treasure_room.exits = {"south": "神秘图书馆"}
        treasure_room.treasure = True
        treasure_room.puzzle_question = "什么东西越洗越脏？"
        treasure_room.puzzle_answer = "2"
        treasure_room.items = [Item("钥匙", "一把神秘的钥匙，可以打开某些门。")]
        
        garden = Room("花园", "一个美丽的花园，充满了奇异的植物和花朵。")
        garden.exits = {"west": "战斗大厅", "east": "地下密室"}
        garden.items = [Item("草药", "可以恢复少量生命值的草药。", "heal", 20)]
        
        dungeon = Room("地下密室", "阴暗潮湿的密室，你听到了奇怪的声音。")
        dungeon.exits = {"west": "花园", "north": "最终宝藏室"}
        dungeon.monster = Monster("骷髅战士", 25, 6)
        
        final_treasure = Room("最终宝藏室", "这是迷宫的最终宝藏室！巨大的宝藏就在眼前！")
        final_treasure.exits = {"south": "地下密室"}
        final_treasure.treasure = True
        
        self.rooms = {
            "入口大厅": entrance,
            "神秘图书馆": library,
            "战斗大厅": battle_room,
            "宝藏室": treasure_room,
            "花园": garden,
            "地下密室": dungeon,
            "最终宝藏室": final_treasure
        }
        
        self.current_room = entrance
        
    def start_game(self):
        print("欢迎来到神秘迷宫寻宝游戏！")
        print("=" * 50)
        print("你将扮演一个勇敢的冒险者，探索神秘的迷宫并寻找宝藏！")
        self.player = Player("冒险者")
        print(f"你好，勇敢的 {self.player.name}！")
        print("你的目标是探索迷宫，收集宝藏，并安全地逃离。")
        print(f"你拥有 {self.player.hp} 点生命值。")
        print("=" * 50)
        
        self.game_loop()
        
    def clear_screen(self):
        os.system('clear' if os.name == 'posix' else 'cls')
        
    def show_room_info(self):
        room = self.current_room
        print(f"\n=== {room.name} ===")
        print(room.description)
        
        if room.exits:
            print(f"出口方向: {', '.join(room.exits.keys())}")
            
        if room.items:
            print("房间内的物品:")
            for item in room.items:
                print(f"- {item.name}: {item.description}")
                
        if room.monster:
            print(f"危险！你发现了 {room.monster.name}！")
            print(f"它拥有 {room.monster.hp} 点生命值，攻击力为 {room.monster.attack}。")
            
        if room.treasure and not room.puzzle_solved:
            print("这里有一道谜题需要解决才能获得宝藏。")
            
        if room.treasure and room.puzzle_solved:
            print("恭喜！你发现了宝藏！")
            
    def show_player_status(self):
        print(f"\n=== {self.player.name} 的状态 ===")
        print(f"生命值: {self.player.hp}/{self.player.max_hp}")
        print(f"攻击力: {self.player.attack}")
        print(f"防御力: {self.player.defense}")
        print(f"收集的宝藏: {self.player.treasures_found}")
        
        if self.player.inventory:
            print("背包中的物品:")
            for item in self.player.inventory:
                print(f"- {item.name}")
                
    def move_player(self, direction):
        if direction in self.current_room.exits:
            new_room_name = self.current_room.exits[direction]
            self.current_room = self.rooms[new_room_name]
            print(f"你移动到了 {self.current_room.name}。")
        else:
            print("那个方向没有出口！")
            
    def pick_item(self):
        room = self.current_room
        if not room.items:
            print("这个房间里没有物品可以拾取。")
            return
            
        print("你想拾取哪个物品？")
        for i, item in enumerate(room.items):
            print(f"{i + 1}. {item.name}")
            
        try:
            choice = int(input("请输入选择: ")) - 1
            if 0 <= choice < len(room.items):
                item = room.items.pop(choice)
                self.player.add_item(item)
                print(f"你拾取了 {item.name}。")
                
                if item.effect_type == "attack":
                    print(f"你的攻击力增加了 {item.effect_value} 点！")
                elif item.effect_type == "defense":
                    print(f"你的防御力增加了 {item.effect_value} 点！")
                elif item.effect_type == "heal":
                    print(f"你恢复了 {item.effect_value} 点生命值！")
            else:
                print("无效的选择。")
        except ValueError:
            print("请输入有效的数字。")
            
    def combat(self):
        room = self.current_room
        if not room.monster:
            print("这里没有怪物可以战斗。")
            return
            
        monster = room.monster
        print(f"战斗开始！你与 {monster.name} 展开了战斗！")
        
        while self.player.hp > 0 and monster.hp > 0:
            # 玩家攻击
            player_damage = random.randint(1, self.player.attack)
            monster.hp -= player_damage
            print(f"你对 {monster.name} 造成了 {player_damage} 点伤害！")
            
            if monster.hp <= 0:
                print(f"你击败了 {monster.name}！")
                room.monster = None
                break
                
            # 怪物攻击
            monster_damage = monster.attack - self.player.defense
            monster_damage = max(1, monster_damage)
            actual_damage = self.player.take_damage(monster_damage)
            print(f"{monster.name} 对你造成了 {actual_damage} 点伤害！")
            
            if self.player.hp <= 0:
                print(f"你被 {monster.name} 击败了...")
                self.game_over = True
                return
                
            time.sleep(1)
            
    def solve_puzzle(self):
        room = self.current_room
        if not room.treasure or room.puzzle_solved:
            print("这里没有谜题需要解决。")
            return
            
        print("你发现了一个神秘的谜题...")
        print(f"谜题：{room.puzzle_question}")
        print("1. 水  2. 手  3. 布  4. 答案是2")
        
        answer = input("请输入你的选择 (1-4): ")
        
        if answer == room.puzzle_answer:
            print("恭喜！谜题解决了！你获得了宝藏！")
            room.puzzle_solved = True
            self.player.treasures_found += 1
        else:
            print("答案错误，再想想吧！")
            
    def show_help(self):
        print("\n=== 游戏帮助 ===")
        print("look - 查看当前房间信息")
        print("status - 查看你的状态")
        print("move [方向] - 移动到指定方向 (north, south, east, west)")
        print("take - 拾取物品")
        print("fight - 战斗")
        print("puzzle - 解决谜题")
        print("inventory - 查看背包")
        print("help - 显示帮助")
        print("quit - 退出游戏")
        
    def game_loop(self):
        print("\n游戏开始！输入 'help' 查看可用命令。")
        
        while not self.game_over:
            print("\n> ", end="")
            command = input().strip().lower()
            
            if command == "quit":
                print("感谢游玩神秘迷宫寻宝游戏！")
                break
            elif command == "help":
                self.show_help()
            elif command == "look":
                self.show_room_info()
            elif command == "status":
                self.show_player_status()
            elif command == "inventory":
                self.show_player_status()
            elif command == "take":
                self.pick_item()
            elif command == "fight":
                self.combat()
            elif command == "puzzle":
                self.solve_puzzle()
            elif command.startswith("move "):
                direction = command[5:]
                self.move_player(direction)
            else:
                print("无效的命令。输入 'help' 查看可用命令。")
                
            # 检查游戏结束条件
            if self.player.treasures_found >= 2:
                print("\n恭喜！你收集了足够的宝藏，成功逃离了迷宫！")
                print("你是一个真正的冒险者！")
                break
                
            if self.player.hp <= 0:
                print("\n游戏结束！你被击败了...")
                break

def main():
    game = MazeGame()
    game.start_game()

if __name__ == "__main__":
    main()